### PR TITLE
fix(ecr-assets): fix loading image tarballs with existing tags

### LINKED ISSUE #18822: (cdk-assets): TarballImageAsset fails build if trying to load tarball with existing tag
### What is the problem?

When loading an image tarball that has different contents, but the same image tag, it will fail to load with an error similar to the following:

```
Error parsing reference: "The image bazel/projects/my_image:my_image already exists, renaming the old one with ID sha256:b9bf91fbc75ee3d7f9d78c7c2f27d06205f8c71f274b444ebfe573c9d159c26f to empty string\nbazel/projects/my_image:my_image" is not a valid repository/tag: invalid reference format: repository name must be lowercase
```

### Reproduction Steps

This is a complete example script to show reproduce the root cause error without a bunch of CDK code:

```
echo "FROM busybox\nWORKDIR /" | docker build -t cdk_example:test -
docker save cdk_example:test -o test.tar
echo "FROM busybox\nWORKDIR /different" | docker build -t cdk_example:test -
docker load -i test.tar
```

When executing that final load, you'll get the output:

```
The image cdk_example:test already exists, renaming the old one with ID sha256:3e782920eed15d9f99ae20b58d734022aacd6f510a128c3baad4377b3541b419 to empty string
Loaded image: cdk_example:test
```

When CDK itself loads an image, it gets the repository and tag for an image by piping the output of `docker load` to `sed "s/Loaded image: //g"`

If that `sed` command is run on the typical output, you correctly get the image name. But when run on the output above, you get:

```
The image cdk_example:test already exists, renaming the old one with ID sha256:3e782920eed15d9f99ae20b58d734022aacd6f510a128c3baad4377b3541b419 to empty string
cdk_example:test
```

Which then fails when CDK attempts to re-tag the image for upload to ECR.

### What did you expect to happen?

The image would deploy successfully

### What actually happened?

The image fails to deploy when trying to load the tarball for pushing

### CDK CLI Version

2.9.0 (build a69725c)

### Framework Version

_No response_

### Node.js Version

v16.13.2

### OS

Linux

### Language

Python

### Language Version

_No response_

### Other information

The following `sed` command is able to accurately extract the image name:

```
sed -nr 's/^Loaded image: (.*)$/\1/p'
```

Credit to this comment for a more robust way to do this:  https://github.com/docker/cli/issues/2212#issuecomment-953224385

### PR DESCRIPTION
**Pull request by AndrewGuenther**
The current logic for pushing tarball images fails if the tarball being
loaded uses a repository and tag that's already loaded into the docker
daemon. This is due to the way the ecr-assets module parses the output
of the docker load command. If the repository/tag combination already
exists in the daemon, it outputs a message about renaming it, which
breaks the `sed` command parsing its output.

This change updates the `sed` command used for extracting the
repository/tag for tarball images to make it more robust and will now
successfully parse docker load output in the case where the
repository/tag already exist.

fixes #18822

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by madeline-k — COMMENTED** [review]
Thanks for opening this PR, @AndrewGuenther! Do you have any ideas on how to test this a bit more thoroughly?

**Comment by AndrewGuenther** [comment]
So the tarball asset code isn't really tested much at all. The test I updated is the only one currently in place. We could add an integration test which specifically tests tarball loading, but the specific case covered here would require an image to already be loaded in the Docker daemon. I'm not sure what the best way to go about that is. Make a subprocess call during test setup?

I'd be happy to improve the overall test coverage of tarball functionality in a separate PR. The CDK testing setup is pretty complex and I've had major issues running it locally (sorry for all the failed CI revisions). So it would be a bit before I could really put my head down and improve that. I'd appreciate if we could land this seeing as it updates the existing tests and consider the general coverage improvements separately.

**Comment by AndrewGuenther** [comment]
@madeline-k Can we get this landed? It's a pretty small change and would help remove a lot of workaround cruft for my org.

File: packages/@aws-cdk/aws-ecr-assets/lib/tarball-asset.ts [hunk 1]
```diff
@@ -78,7 +78,7 @@ export class TarballImageAsset extends CoreConstruct implements IAsset {
       executable: [
         'sh',
         '-c',
-        `docker load -i ${relativePathInOutDir} | sed "s/Loaded image: //g"`,
+        `docker load -i ${relativePathInOutDir} | sed -nr 's/^Loaded image: (.*)$/\\1/p'`,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecr-assets/lib/tarball-asset.ts:81]: This code will not work on MacOS, as `-r` is not a valid flag for `sed` on that platform.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-ecr-assets/lib/tarball-asset.ts (hunk 1 above)
Comment (AndrewGuenther) [reply packages/@aws-cdk/aws-ecr-assets/lib/tarball-asset.ts:81]: Well that's a bit of a piss off...hmmmm. I'll work on figuring our an alternative. Technically what's really needed is to strip out any following lines. It isn't my ideal, but we could pipe this into `head` which should work better cross platform...

Comment (AndrewGuenther) [reply packages/@aws-cdk/aws-ecr-assets/lib/tarball-asset.ts:81]: @rix0rrr For a test like that to work, I would need to communicate with the docker daemon directly in order to replicate the issue being fixed here. Is there an existing precedent for running shell commands in integ tests? I had pushed back earlier when @madeline-k suggested this as I don't think that this PR is the right place to backfill testing for a completely untested class.

**Comment by AndrewGuenther** [comment]
Will leaving a comment keep this open if I'm waiting on follow-up comments from reviewers? A bit frustrating that it takes so long to get a question answered that a bot will close my PR...

**Comment by AndrewGuenther** [comment]
Are you serious?

**Comment by skinny85** [comment]
@AndrewGuenther apologies, our bots can be a little ham-fisted sometimes 🙂.

@rix0rrr can you reply to Andrew's comment in https://github.com/aws/aws-cdk/pull/18823#discussion_r838804900?

**Comment by AndrewGuenther** [comment]
Thanks @skinny85!

**Pull request closed without merging**